zerver/lib/encryption.py

```diff
@@ -0,0 +1,52 @@
+import base64
+import os
+from typing import Tuple
+
+from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+
+from zerver.models import PushDeviceToken
+
+
+def bytes_to_b64(data: bytes) -> str:
+    """
+    Converts bytes to url-safe base64 encoded string.
+    """
+    return base64.b64encode(data).decode('utf-8')
+
+def b64_to_bytes(data: str) -> bytes:
+    """
+    Converts url-safe base64 encoded string to bytes.
+    """
+    return base64.b64decode(data.encode('utf-8'))
+
+def encrypt_data(key: str, data: str) -> Tuple[str, str]:
+    """
+    Returns a tuple of `(encrypted_data, nonce)`.
+    """
+    # AESGCM() expects key as bytes.
+    aesgcm = AESGCM(b64_to_bytes(key))
+    nonce = os.urandom(12)
+    # AESGCM.encrypt() expects data in bytes and returns ciphertext
+    # bytes with the 16 byte tag appended.
+    encrypted_data = aesgcm.encrypt(nonce, data.encode('utf-8'), None)
+    return bytes_to_b64(encrypted_data), bytes_to_b64(nonce)
```

Comment:
We also need to send this information i.e. `encrypted_data` and `nonce` to other devices via bouncer. Sending these fields as bytes will cause the following exception when we try to covert `encrypted_payloads` to JSON format.
```
TypeError: Type is not JSON serializable: bytes
```

Am I missing something?

Reply:
Right. At the moment when we're sending them over the network, they need to be encoded somehow, and base64 is a fine choice there. But the best point to do that encoding is just before we send them over the network, so that inside our code they can have their natural form, which is a plain `bytes`.